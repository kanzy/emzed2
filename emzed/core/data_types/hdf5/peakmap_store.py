# encoding: utf-8
from __future__ import print_function, division

import itertools

from tables import Atom, UInt32Col, Float32Col, UInt8Col, UInt64Col, StringCol
import numpy as np

from emzed_optimizations import sample_peaks as optim_sample_peaks

from .. import PeakMap, Spectrum

from .store_base import Store, filters
from .lru import LruDict, lru_cache

from .install_profile import profile


class PeakMapStore(Store):

    ID_FLAG = 0
    HANDLES = PeakMap

    MSLEVEL_FIELD_SIZE = 16

    def __init__(self, file_, node, **kw):
        self.file_ = file_
        self.node = node
        self.setup()
        self.write_cache = LruDict(100)
        self.read_cache = LruDict(100)

    def setup(self):
        self.setup_blobs()
        self.setup_spec_table()
        self.setup_peakmap_table()

    def setup_blobs(self):
        if not hasattr(self.node, "mz_blob"):
            self.file_.create_earray(self.node, "mz_blob",
                                     Atom.from_dtype(np.dtype("float64")), (0,),
                                     filters=filters,
                                     )

        if not hasattr(self.node, "ii_blob"):
            self.file_.create_earray(self.node, "ii_blob",
                                     Atom.from_dtype(np.dtype("float32")), (0,),
                                     filters=filters,
                                     )

    def setup_spec_table(self):
        if not hasattr(self.node, "spec_table"):
            description = {}
            description["pm_index"] = UInt32Col(pos=0)
            description["rt"] = Float32Col(pos=1)
            description["ms_level"] = UInt8Col(pos=2)
            description["start"] = UInt64Col(pos=3)
            description["size"] = UInt32Col(pos=4)
            spec_table = self.file_.create_table(self.node, 'spec_table', description,
                                                 filters=filters)

            # every colums which appears in a where method call should/must be indexed !
            # this is not only for performance but for correct lookup as well (I had strange bugs
            # else)
            spec_table.cols.pm_index.create_index()
            spec_table.cols.rt.create_index()

    def setup_peakmap_table(self):
        if not hasattr(self.node, "pm_table"):
            description = {}
            description["unique_id"] = StringCol(itemsize=64, pos=0)
            description["index"] = UInt32Col(pos=1)
            description["ms_levels"] = StringCol(itemsize=self.MSLEVEL_FIELD_SIZE, pos=2)
            description["rtmin"] = Float32Col(pos=3)
            description["rtmax"] = Float32Col(pos=4)
            description["mzmin"] = Float32Col(pos=5)
            description["mzmax"] = Float32Col(pos=6)
            pm_table = self.file_.create_table(self.node, 'pm_table', description, filters=filters)
            # every colums which appears in a where method call should/must be indexed !
            # this is not only for performance but for correct lookup as well (I had strange bugs
            # else)
            pm_table.cols.unique_id.create_index()
            pm_table.cols.index.create_index()

    def add_spectrum(self, pm_index, spec):

        start = self.node.mz_blob.nrows
        self.node.mz_blob.append(spec.peaks[:, 0])
        self.node.ii_blob.append(spec.peaks[:, 1])

        size = spec.peaks.shape[0]

        row = self.node.spec_table.row
        row["pm_index"] = pm_index
        row["rt"] = spec.rt
        row["ms_level"] = spec.msLevel
        row["start"] = start
        row["size"] = size

        row.append()

    def _write(self, col_index, pm):

        # at the moment we ignore col_index, I guess it would not speed up so much

        unique_id = pm.uniqueId()
        # hash key
        yield unique_id

        result = list(self.node.pm_table.where("""unique_id == %r""" % unique_id))
        if result:
            yield result[0]["index"]
            return

        ms_levels = sorted(set(pm.getMsLevels()))
        as_str = ",".join(map(str, ms_levels))
        if len(as_str) > self.MSLEVEL_FIELD_SIZE:
            raise ValueError("ms level description %r is to long" % as_str)

        rtmin, rtmax = pm.rtRange()
        mzmin, mzmax = pm.mzRange()

        index = self.node.pm_table.nrows
        row = self.node.pm_table.row
        row["unique_id"] = unique_id
        row["index"] = index
        row["ms_levels"] = as_str
        row["rtmin"] = rtmin
        row["rtmax"] = rtmax
        row["mzmin"] = mzmin
        row["mzmax"] = mzmax
        row.append()

        for spec in pm.spectra:
            self.add_spectrum(index, spec)

        yield index

    def _read(self, col_index, index):
        result = list(self.node.pm_table.where("""index == %r""" % index))
        if len(result) == 0:
            raise ValueError("index %d not in table" % index)

        ms_levels = result[0]["ms_levels"]
        rtmin = result[0]["rtmin"]
        rtmax = result[0]["rtmax"]
        mzmin = result[0]["mzmin"]
        mzmax = result[0]["mzmax"]
        unique_id = result[0]["unique_id"]
        ms_levels = map(int, ms_levels.split(","))

        result = PeakMapProxy(node=self.node,
                              index=index,
                              ms_levels=ms_levels,
                              rtmin=rtmin,
                              rtmax=rtmax,
                              mzmin=mzmin,
                              mzmax=mzmax,
                              unique_id=unique_id)
        return result

    def flush(self):
        self.node.pm_table.flush()
        self.node.spec_table.flush()


class PeakMapProxy(object):

    def __init__(self, **kw):
        for name, value in kw.items():
            setattr(self, name, value)

        self.meta = {"unique_id": kw.get("unique_id")}

    def uniqueId(self):
        return self.unique_id

    def getMsLevels(self):
        return self.ms_levels

    def rtRange(self):
        return self.rtmin, self.rtmax

    def mzRange(self):
        return self.mzmin, self.mzmax

    def _iter_peaks(self, rtmin, rtmax, mzmin, mzmax, ms_level=1):
        if rtmin is not None and rtmax is not None:
            query = """(pm_index == %d) & (%f <= rt) & (rt <= %f)""" % (self.index, rtmin, rtmax)
        elif rtmin is not None:
            query = """(pm_index == %d) & (%f <= rt)""" % (self.index, rtmin)
        elif rtmax is not None:
            query = """(pm_index == %d) & (rt <= %f)""" % (self.index, rtmax)
        else:
            query = """(pm_index == %d)""" % (self.index,)
        mz_blob = self.node.mz_blob
        ii_blob = self.node.ii_blob
        for row in self.node.spec_table.where(query):
            if ms_level != row["ms_level"]:
                continue
            rt = row["rt"]
            i_start = row["start"]
            i_end = i_start + row["size"]
            mzs = mz_blob[i_start:i_end]
            iis = ii_blob[i_start:i_end]

            flags = True
            if mzmin is not None:
                flags = flags * (mzmin <= mzs)
            if mzmax is not None:
                flags = flags * (mzs <= mzmax)

            if flags is not True:
                idx = np.where(flags)[0]
                iis = iis[idx]
                mzs = mzs[idx]

            yield rt, mzs, iis

    @lru_cache(maxsize=1000)
    def chromatogram(self, mzmin=None, mzmax=None, rtmin=None, rtmax=None, ms_level=1):
        rts = []
        intensities = []
        for rt, mzs, iis in self._iter_peaks(rtmin, rtmax, mzmin, mzmax, ms_level):
            rts.append(rt)
            intensities.append(np.sum(iis))
        rts = np.array(rts)
        intensities = np.array(intensities)
        perm = np.argsort(rts)
        rts = rts[perm]
        intensities = intensities[perm]
        return rts, intensities

    @lru_cache(maxsize=1000)
    def sample_peaks(self, rtmin, rtmax, mzmin, mzmax, npeaks, ms_level):
        spectra = []
        for rt, mzs, iis in self._iter_peaks(rtmin, rtmax, mzmin, mzmax, ms_level):
            peaks = np.vstack((mzs, iis)).T
            spectrum = Spectrum(peaks, rt, ms_level, "0", [])
            spectra.append(spectrum)
        if rtmin <= rtmax and mzmin < mzmax:
            peaks = optim_sample_peaks(PeakMap(spectra),
                                       rtmin, rtmax, mzmin, mzmax, npeaks, ms_level)
        else:
            peaks = np.zeros((0, 2))
        return peaks

    def get_rts(self, msLevel=1):
        t = self.node.spec_table
        pm_indices = t.cols.pm_index[:]
        rts = t.cols.rt[:]
        ms_levels = t.cols.ms_level[:]
        return [r for (r, i, l) in itertools.izip(rts, pm_indices, ms_levels) if i == self.index and
                l == msLevel]
