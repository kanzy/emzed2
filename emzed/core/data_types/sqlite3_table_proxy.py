# encoding: utf-8
from __future__ import print_function, division, absolute_import

import operator
import sqlite3


from .sqlite_helpers import unpickle_object, chromatogram


class Sqlite3TableProxy(object):

    def __init__(self, path):
        self.conn = sqlite3.connect(path)
        self.conn.execute('''ATTACH DATABASE ':memory:' AS mem''')
        self.read_table_info()

        self.sort_order = ""
        self.select_for_mixed_sort_order = None
        self.reset_filter()
        self.cursor = None

    def read_table_info(self):
        __, info = unpickle_object(1, self.conn)

        for (name, value) in info._asdict().items():
            setattr(self, name, value)

    def reset_filter(self):
        self.filter_expression = "0=0"

    def create_query(self):
        stmt = "SELECT * FROM rows WHERE {filter_} {order}".format(filter_=self.filter_expression,
                                                                   order=self.sort_order)
        if self.cursor is not None:
            self.cursor.close()
        print(stmt)
        self.cursor = self.conn.cursor().execute(stmt)
        return stmt

    def fetchone(self):
        assert self.cursor is not None, "call create_query() method first"
        return self.cursor.fetchone()

    def set_sort_order(self, items):
        assert all(len(item) == 2 for item in items)

        if len(items) == 0:
            self.sort_order = ""
            return

        term = ", ".join("%s %s" % (name, "ASC" if ascending else "DESC") for (name, ascending)
                         in items)
        self.sort_order = "ORDER BY %s" % term

    def set_filter(self, filter_expression):
        self.filter_expression = filter_expression

    def check_filter_expression(self, filter_expression):
        stmt = "SELECT * FROM rows WHERE (%s) LIMIT 1" % filter_expression
        cursor = self.conn.cursor()
        try:
            cursor.execute(stmt)
        except Exception as e:
            return str(e)
        return None

    def get_chromatogram(self, rtmin, rtmax, mzmin, mzmax, peakmap_entry):
        peakmap_id = int(peakmap_entry.split("!")[0])
        return chromatogram(self.conn, peakmap_id, rtmin, rtmax, mzmin, mzmax)
