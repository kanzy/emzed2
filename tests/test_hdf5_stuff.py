# encoding: utf-8, division
from __future__ import print_function, division

import os
import tempfile

from datetime import datetime

import numpy as np

from emzed.core import TimeSeries, CheckState, PeakMap, Spectrum
from emzed.io import to_hdf5, append_to_hdf5, atomic_hdf5_writer

from emzed.core.data_types.hdf5_table_proxy import Hdf5TableProxy, ObjectProxy, PeakMapProxy

from emzed.core.data_types.hdf5.accessors import (Hdf5TableWriter, Hdf5TableAppender,
                                                  Hdf5TableReader)

from emzed.core.data_types.hdf5.bit_matrix import (BitMatrix,)


from emzed.utils import toTable

from pytest import fixture, raises, yield_fixture
import pytest


@fixture()
def table():

    peaks_0 = np.arange(20.0).reshape(-1, 2)
    peaks_1 = np.arange(30.0).reshape(-1, 2)

    spec_0 = Spectrum(peaks_0[:], 30.0, 1, "+")
    spec_1 = Spectrum(peaks_1[:], 30.0, 2, "+", [(10, 1000)])
    spec_2 = Spectrum(peaks_1[:], 40.0, 1, "+")

    pm = PeakMap([spec_0, spec_1, spec_2])

    t0 = toTable("int", (None, 2, 3, 3, None), type_=int)
    t0.addColumn("float", (1.0, 2.0, 4.0, 3.0, None), type_=float)
    t0.addColumn("bool", (True, False, None, True, False), type_=bool)
    t0.addColumn("check", (True, False, None, True, False), type_=CheckState)
    t0.addColumn("str", ("1", "2" * 100, None, "a", "b"), type_=str)
    t0.addColumn("object", ({1}, dict(a=2), None, (1,), [1, 2]), type_=object)
    t0.addColumn("peakmap", pm, type_=PeakMap)
    t0.addColumn("i1", [1, None, 2, None, 3], type_=int)
    t0.addColumn("i2", [None, 1, None, 2, None], type_=int)

    n = 10
    ts_0 = TimeSeries(map(datetime.fromordinal, range(1, n + 1)), range(n), "label1")
    ts_0.x[3] = None
    ts_0.y[3] = np.nan

    n = 10
    ts_1 = TimeSeries(map(datetime.fromordinal, range(1000,  n + 1000)),
                      range(1000, 1000 + n), "label2", [bool(i % 4) for i in range(n)])

    t0.addColumn("time_series", [ts_0, ts_0, ts_1, ts_1, None], type_=TimeSeries)

    t0.meta = {"a": "b"}
    return t0


@yield_fixture()
def tproxy(table):

    folder = tempfile.mkdtemp()
    path = os.path.join(folder, "t.hdf5")
    to_hdf5(table, path)

    tproxy = Hdf5TableProxy(path)
    yield tproxy
    tproxy.close()


def test_writer_appender_reader(table, tmpdir, regtest):

    path = tmpdir.join("test.hdf5").strpath
    table = table.extractColumns("int", "str", "check", "object", "peakmap")
    writer = Hdf5TableWriter(path)
    writer.write_table(table)
    writer.close()

    appender = Hdf5TableAppender(path)
    appender.append_table(table)
    appender.append_table(table)
    appender.close()

    reader = Hdf5TableReader(path)
    assert len(reader) == 3 * len(table)
    rows = list(reader)

    # resolve proxies:
    rows = [[ci.load() if isinstance(ci, ObjectProxy) else ci for ci in row] for row in rows]

    # check if appending worked:
    assert rows[:5] == rows[5:10]
    assert rows[5:10] == rows[10:]

    for col in reader.col_names:
        print(reader.get_col_values(col), file=regtest)

    reader.close()


def test_hdf5_table_appender(table, tmpdir):
    path = tmpdir.join("test.hdf5").strpath
    to_hdf5(table, path)
    append_to_hdf5([table, table, table], path)


def test_atomic_hdf5_writer(table, tmpdir, regtest):
    path = tmpdir.join("test.hdf5").strpath
    with atomic_hdf5_writer(path) as add:
        add(table)
        table.replaceColumn("int", table.int + 10, type_=int)
        add(table)
        table.replaceColumn("int", table.int + 10, type_=int)
        add(table)
        table.replaceColumn("int", table.int + 10, type_=int)

    t = Hdf5TableProxy(path).toTable()
    print(t.int.values, file=regtest)


def test_round_trip(tproxy, table, regtest):
    # test roundtrip:
    t1 = tproxy.toTable()
    assert t1.uniqueId() == table.uniqueId()
    assert t1.meta == table.meta

    for pm0, pm1 in zip(t1.peakmap, table.peakmap):
        assert isinstance(pm0, PeakMapProxy)
        assert isinstance(pm1, PeakMap)
        rt0, ii0 = pm0.chromatogram(0, 1000, 35, 55)   # fetch lazy
        rt1, ii1 = pm1.chromatogram(0, 1000, 35, 55)   # data is in memory
        assert np.all(rt0 == rt1)
        assert np.all(ii0 == ii1)
        rt0, ii0 = pm0.chromatogram(0, 1000, 0, 1000)   # fetch lazy
        rt1, ii1 = pm1.chromatogram(0, 1000, 0, 1000)   # data is in memory
        assert np.all(rt0 == rt1)
        assert np.all(ii0 == ii1)
        rt0, ii0 = pm0.chromatogram(0, 1000, 1000, 1100)   # fetch lazy
        rt1, ii1 = pm1.chromatogram(0, 1000, 1000, 1200)   # data is in memory
        assert np.all(rt0 == rt1)
        assert np.all(ii0 == ii1)

    print(t1, file=regtest)
    return


def test_round_trip_objects(tmpdir, regtest):
    # test roundtrip:
    col = [{i: i + 1} for i in range(5)]
    t0 = toTable("dicts", col, type_=object)

    col = [{i: i + 2} for i in range(5)]
    t0.addColumn("dicts_1", col, type_=object)
    t0.addColumn("strs", map(str, range(5)), type_=str)
    t0.addColumn("strs_1", t0.strs * 3 + "_x", type_=str)

    print(t0, file=regtest)
    path = tmpdir.join("test.hdf5").strpath
    to_hdf5(t0, path)
    tproxy = Hdf5TableProxy(path)

    tback = tproxy.toTable()
    print(tback, file=regtest)


def test_get_index(tproxy):

    # test getIndex method
    assert tproxy.getIndex("int") == 0
    assert tproxy.getIndex("object") == 5

    with raises(Exception):
        tproxy.getIndex("hih")


def test_row_attribute(tproxy, regtest):
    # test row attribute
    for row in tproxy.rows:
        print(row, file=regtest)


def test_row_write_cell(tproxy, regtest):
    # test row attribute
    tproxy.setCellValue(0, 0, 4711)

    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    # make sure that row 0 is cached, and test again:
    tproxy[0]
    tproxy.setCellValue(0, 0, 4712)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(4, 0, 4713)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(4, 0, None)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(4, 0, 4713)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(4, 0, None)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(0, 3, "4713")
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(0, 3, None)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)


def test_sort_by(tproxy, regtest):
    # test sortBy
    perm = tproxy.sortPermutation("int", True)
    print(perm, file=regtest)
    print(tproxy.toTable().extractColumns("int")[perm], file=regtest)

    perm = tproxy.sortPermutation("int", False)
    print(perm, file=regtest)
    print(tproxy.toTable()[perm].extractColumns("int"), file=regtest)

    perm = tproxy.sortPermutation(("int", "float"), (True, True))
    print(perm, file=regtest)
    print(tproxy.toTable()[perm].extractColumns("int", "float"), file=regtest)

    perm = tproxy.sortPermutation(("int", "float"), (True, False))
    print(perm, file=regtest)
    print(tproxy.toTable()[perm].extractColumns("int", "float"), file=regtest)


def test_replace_column(tproxy, regtest):

    tproxy.replaceColumn("int", 2)
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("int", None)
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("int", 1)
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("int", 3)
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("int", None)
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("int", 3)
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)


def test_replace_column_2(tproxy, regtest):
    tproxy.replaceColumn("str", "2")
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("str", None)
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("str", "1")
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("str", "3")
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("str", None)
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceColumn("str", "3")
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)


def test_replace_column_3(tproxy, regtest):

    tproxy.replaceSelectedRows("int", 2, range(5))
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceSelectedRows("int", None, range(5))
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceSelectedRows("int", 1, range(5))
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceSelectedRows("int", 3, range(5))
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceSelectedRows("int", None, range(5))
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)

    tproxy.replaceSelectedRows("int", 3, range(5))
    print(tproxy.toTable().extractColumns("int", "float", "bool", "str"), file=regtest)


def test_selected_col_values(tproxy, regtest):
    for n in (1, 2, 3, 4, 5):
        v = tproxy.selectedRowValues("int", range(n))
        print(v, file=regtest)

    for n in (1, 2, 3, 4):
        v = tproxy.selectedRowValues("int", range(n, 5))
        print(v, file=regtest)


def test_ghost_table_forwarding(tproxy):
    assert tproxy.getValue(tproxy.rows[1], "int") == 2

    assert tproxy.supportedPostfixes("") == []


@pytest.mark.skip(reason="not fixed yet")
def test_version_2_26_0(path, regtest):
    t = Hdf5TableProxy(path("data/table_v_2_26_0.hdf5"))
    assert t.hdf5_meta == {'hdf5_table_version': (2, 26, 0)}

    # calling .toTable is the "lackmus" test if data reading worked:
    print(t.toTable(), file=regtest)


def test_bit_matrix(tmpdir):
    from tables import open_file
    for n in (11, 53):
        file_ = open_file(tmpdir.join("data.bin").strpath, "w")
        bm = BitMatrix(file_, "flags", n)
        bm.resize(5)
        for row in range(5):
            for col in range(n):
                bm.set_bit(row, col)
                assert row in bm.positions_in_col(col)
                assert col in bm.positions_in_row(row)
                bm.unset_bit(row, col)
        file_.close()


@pytest.fixture
def proxy_small_table(tmpdir):
    t = toTable("a", (1, 1, 2, 2, None), type_=int)
    t.addColumn("d", ("4", "4", "2", "1", None), type_=str)
    path = tmpdir.join("test.hdf5").strpath
    to_hdf5(t, path)

    prox = Hdf5TableProxy(path)
    return prox


def test_perm(regtest, proxy_small_table):

    prox = proxy_small_table

    t = prox.toTable()

    perm = prox.sortPermutation(("a", "d"), (True, True))
    print("ascending, ascending", file=regtest)
    print(perm, file=regtest)
    print(t[perm], file=regtest)

    perm = prox.sortPermutation(("a", "d"), (False, True))
    print("descending, ascending", file=regtest)
    print(perm, file=regtest)
    print(t[perm], file=regtest)

    perm = prox.sortPermutation(("a", "d"), (True, False))
    print("ascending, descending", file=regtest)
    print(perm, file=regtest)
    print(t[perm], file=regtest)

    perm = prox.sortPermutation(("a", "d"), (False, False))
    print("descending, descending", file=regtest)
    print(perm, file=regtest)
    print(t[perm], file=regtest)

    perm = prox.sortPermutation("a", True)
    print("ascending", file=regtest)
    print(perm, file=regtest)
    print(t[perm], file=regtest)

    perm = prox.sortPermutation("a", False)
    print("descending", file=regtest)
    print(perm, file=regtest)
    print(t[perm], file=regtest)


def test_csv_write(regtest, tmpdir, proxy_small_table):

    csv_path = tmpdir.join("test.csv").strpath
    proxy_small_table.storeCSV(csv_path)
    regtest.write(open(csv_path, "r").read())

    proxy_small_table.storeCSV(csv_path, row_indices=(1, 3))
    regtest.write(open(csv_path, "r").read())


def test_rows_access(regtest, proxy_small_table):

    for row in proxy_small_table.rows:
        print(row, file=regtest)


def test_write_unicode(tmpdir, regtest):

    table = toTable("a", ("x", None, u"", "", u"ccccäää##c"), type_=unicode)
    table.addColumn("b", ("x", "asbcder", None, None, ""), type_=str)
    table.addColumn("c", ("x", "asbcder", 1, (1,), ""), type_=object)
    path = tmpdir.join("t.hdf5").strpath
    to_hdf5(table, path)
    prox = Hdf5TableProxy(path)
    print(prox.toTable(), file=regtest)
    print(prox.reader.get_col_values("a"), file=regtest)
    print(prox.reader.get_col_values("b"), file=regtest)
    print([i.load() for i in prox.reader.get_col_values("c")], file=regtest)


def test_unique_col_value(proxy_small_table, regtest):
    values = proxy_small_table.getUniqueValues("a")
    assert sorted(set(values)) == sorted(values)
    assert None not in values
    print(sorted(values), file=regtest)

    values = proxy_small_table.getUniqueValues("d")
    assert sorted(set(values)) == sorted(values)
    assert None not in values
    print(sorted(values), file=regtest)


def test_unique_col_value_extended(tproxy, regtest):
    values = tproxy.getUniqueValues("time_series")
    for v in values:
        assert isinstance(v, TimeSeries)
    ids = [ts.uniqueId() for ts in values]
    print(sorted(ids), file=regtest)

"""
todo:
    blank: auch zukunft
    """
