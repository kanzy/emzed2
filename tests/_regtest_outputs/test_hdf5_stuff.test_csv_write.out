a;d
1;4
1;4
2;2
2;1
-;-
a;d
1;4
2;1
